### PR TITLE
Generate: Mistral/Mixtral FA2 cache fix when going beyond the context window

### LINKED ISSUE #27985: `KeyError: 'Cache only has 0 layers, attempted to access layer with index 0'`
### System Info

- `transformers` version: 4.36.0
- Platform: Linux-5.15.0-70-generic-x86_64-with-glibc2.35
- Python version: 3.11.4
- Huggingface_hub version: 0.19.4
- Safetensors version: 0.3.3
- Accelerate version: 0.25.0.dev0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Yes

### Who can help?

@gante 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, TextStreamer

long_text = # ...

model_id = "mistralai/Mixtral-8x7B-Instruct-v0.1"
tokenizer = AutoTokenizer.from_pretrained(model_id)

model = AutoModelForCausalLM.from_pretrained(model_id, use_flash_attention_2=True, torch_dtype=torch.float16, device_map="auto")

streamer = TextStreamer(tokenizer, skip_prompt=True, skip_special_tokens=True)

messages = [
    {"role": "user", "content": f"Summarize the following:\n{long_text}"}
]

inputs = tokenizer.apply_chat_template(messages, return_tensors="pt").to(model.device)

outputs = model.generate(inputs, max_new_tokens=8192, do_sample=True, streamer=streamer)
print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

### Expected behavior

Expected it to work or at least give me a cuda error.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

The FA2 code path was indexing the `Cache` object incorrectly. This PR fixes it.

Fixes #27985 

_____________________________________________________________

NOTE: `tests/models/mistral/test_modeling_mistral.py::MistralIntegrationTest::test_model_7b_long_prompt` (slow test) was failing on `main`, but it was not popping up in our daily slow CI 🤔 because of that, this issue flew under the radar. It is passing now.

Edit: the test was not run because we are skipping FA2 tests (`@require_flash_attn`). @ydshieh is on it :)

### CODE REVIEW

File: src/transformers/models/mistral/modeling_mistral.py [hunk 1]
```diff
@@ -385,11 +385,16 @@ def forward(
 
         if past_key_value is not None:
```
Comment (gante) [inline src/transformers/models/mistral/modeling_mistral.py:386]: context: when `use_cache is True`, `past_key_value` is now a `Cache` object even if it is an empty cache (previously it was `None`).

As such, with a slicing window, we need to check whether the cache has contents before attempting to slice, as we can't slice `None`.

File: src/transformers/models/mistral/modeling_mistral.py [hunk 2]
```diff
@@ -400,8 +405,6 @@ def forward(
                         f" {past_key.shape}"
                     )
 
-                past_key_value = (past_key, past_value)
```
Comment (gante) [inline src/transformers/models/mistral/modeling_mistral.py:403]: `past_key_value` is now a `Cache` instance that is update in place with the `.update()` function (L413 in the updated file)

File: src/transformers/models/mistral/modeling_mistral.py [hunk 3]
```diff
@@ -385,11 +385,16 @@ def forward(
...
+            if (
+                getattr(self.config, "sliding_window", None) is not None
+                and kv_seq_len > self.config.sliding_window
+                and cache_has_contents
+            ):
                 slicing_tokens = 1 - self.config.sliding_window
 
-                past_key = past_key_value[0]
-                past_value = past_key_value[1]
+                past_key = past_key_value[self.layer_idx][0]
```
Comment (tomaarsen) [inline src/transformers/models/mistral/modeling_mistral.py:396]: `self.layer_idx` is guaranteed to be defined here, right?

**Comment by tomaarsen** [comment]
This indeed seems to address what was described [here](https://github.com/huggingface/transformers/issues/27985#issuecomment-1855524359), well done!

File: src/transformers/models/mistral/modeling_mistral.py (hunk 3 above)
Comment (gante) [reply src/transformers/models/mistral/modeling_mistral.py:396]: Nope, good catch! Going to add an appropriate exception.

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing! 

Looking forward to the FA2 tests being run 😅

**Comment by ydshieh** [comment]
As @younesbelkada mentioined and in the official site

> Ampere, Ada, or Hopper GPUs (e.g., A100, RTX 3090, RTX 4090, H100). Support for Turing GPUs (T4, RTX 2080) is coming soon, please use FlashAttention 1.x for Turing GPUs for now.

nothing we can't do unless we run on  a different machine

**Pull request merged**